skbio/metadata/_interval.py

```diff
@@ -0,0 +1,312 @@
+        >>> interval_metadata.query(intervals=[(1, 2)])
+        [Interval(intervals=[(0, 2), (4, 7)], metadata={'name': 'sagA'})]
+
+        Note
+        ----
+        There are two types of queries to perform
+        1. Query by interval.
+        2. Query by key/val pair (i.e. gene=sagA).
+
+        """
+        if self._is_stale_tree:
+            self._rebuild_tree(self._metadata)
+            self._is_stale_tree = False
+
+        invs = set()
+
+        # Find queries by interval
+        if intervals is not None:
+            for value in intervals:
+                invs.update(self._query_interval(value))
+
+        # Find queries by feature attribute
+        if len(invs) == 0 and metadata is not None:
+            invs = set(self._metadata)
+
+        if metadata is not None:
+            invs = self._query_attribute(list(invs), metadata)
+        return list(invs)
+
+    def drop(self, intervals=None, boundaries=None, metadata=None):
+        """ Drops Interval objects according to a specified query.
+
+        Drops all Interval objects that matches the query.
+
+        Parameters
+        ----------
+        intervals : iterable of tuple of ints
+            A list of intervals associated with the Interval object.
+        boundaries : iterable of tuple of bool
+            A list of boundaries associated with the Interval object.
+        metadata : dict
+            A dictionary of key word attributes associated with the
+            Interval object.
+
+        Examples
+        --------
+        >>> from skbio.metadata import IntervalMetadata
+        >>> interval_metadata = IntervalMetadata()
+        >>> interval_metadata.add(intervals=[(0, 2), (4, 7)],
+        ...                       boundaries=None, metadata={'name': 'sagA'})
+        >>> interval_metadata.add(intervals=[(40, 70)],
+        ...                       boundaries=None, metadata={'name': 'sagB'})
+        >>> interval_metadata.drop(metadata={'name': 'sagA'})
+        >>> interval_metadata.query(metadata={'name': 'sagA'})
+        []
+        """
+        if intervals is None:
+            intervals = []
+        if metadata is None:
+            metadata = {}
+
+        queried_invs = self.query(intervals=intervals,
+                                  boundaries=boundaries,
+                                  metadata=metadata)
+        new_invs = []
+        # iterate through queries and drop them
+        for inv in self._metadata:
+            if inv not in queried_invs:
+                new_invs.append(inv)
+        self._metadata = new_invs
+        self._is_stale_tree = True
+
+    def __eq__(self, other):
+        return sorted(self._metadata) == sorted(other._metadata)
+
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+
+def _polish_interval(interval):
+    if isinstance(interval, tuple):
+        if len(interval) == 0:
+            return None
+        start, end = interval
+        if (len(interval) != 2 or
```

Comment:
Oh ...
On May 19, 2016 11:46 PM, "Zech Xu" <notifications@github.com> wrote:

> In skbio/metadata/_interval.py
> <https://github.com/biocore/scikit-bio/pull/1356#discussion_r63997852>:
>
> > +        self._metadata = new_invs
> > +        self._is_stale_tree = True
> > +
> > +    def __eq__(self, other):
> > +        return sorted(self._metadata) == sorted(other._metadata)
> > +
> > +    def __ne__(self, other):
> > +        return not self.__eq__(other)
> > +
> > +
> > +def _polish_interval(interval):
> > +    if isinstance(interval, tuple):
> > +        if len(interval) == 0:
> > +            return None
> > +        start, end = interval
> > +        if (len(interval) != 2 or
>
> I mean you should check before the unpacking (start, end = interval) not
> after it.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/biocore/scikit-bio/pull/1356/files/4767f8e7b6de74b0308879ee2fc2b7da9dceb048#r63997852>
>